runtime/src/stakes.rs

```diff
@@ -253,31 +287,13 @@ impl Stakes<StakeAccount> {
                 return Err(Error::VoteAccountMismatch(*pubkey));
             }
         }
-        // Assert that all valid vote-accounts referenced in
-        // stake delegations are already cached.
-        let voter_pubkeys: HashSet<Pubkey> = stakes
-            .stake_delegations
-            .values()
-            .map(|delegation| delegation.voter_pubkey)
-            .filter(|voter_pubkey| stakes.vote_accounts.get(voter_pubkey).is_none())
-            .collect();
-        for pubkey in voter_pubkeys {
-            let Some(account) = get_account(&pubkey) else {
-                continue;
-            };
-            if VoteStateVersions::is_correct_size_and_initialized(account.data())
-                && VoteAccount::try_from(account.clone()).is_ok()
-            {
-                error!("vote account not cached: {pubkey}, {account:?}");
-                return Err(Error::VoteAccountNotCached(pubkey));
-            }
-        }
+
         Ok(Self {
-            vote_accounts: stakes.vote_accounts.clone(),
-            stake_delegations: stake_delegations.collect::<Result<_, _>>()?,
+            vote_accounts: stakes.vote_accounts,
+            stake_delegations,
             unused: stakes.unused,
             epoch: stakes.epoch,
-            stake_history: stakes.stake_history.clone(),
```

Comment:
These two `clones` here are pretty cheap and a non-issue.
So no need to require ownership of `stakes: Stakes<Delegation>` argument which will require an expensive clone of `stake_delegations: ImHashMap<Pubkey, Delegation>` field if we can't move the whole thing.

I understand today's code can just move the entire `stakes`, but code is subject to change and the api here does not really need to make that assumption.

Reply:
> These two clones here are pretty cheap and a non-issue.
> So no need to require ownership of stakes: Stakes<Delegation> argument which will require an expensive clone of stake_delegations: ImHashMap<Pubkey, Delegation> field if we can't move the whole thing.

Cloning im::HashMap is not expensive which is kinda the whole point for using it over std HashMap?

> I understand today's code can just move the entire stakes, but code is subject to change and the api here does not really need to make that assumption.

To me this seems like a map (which is conceptually a move) of Stakes<Delegation> to Stakes<StakeAccount> and in fact the caller is immediately dropping Stakes<Delegation> if we don't move it. I can undo this, but seems backwards to me and frankly it feels like you're asking me to do a change for some imaginary code that will never exist.

Reply:
Also note that I can use your argument to argue the other side:

> These two clones here are pretty cheap and a non-issue.

Code is subject to change, we could remove Arcs and clones could become expensive! (I am not actually arguing this to be clear :P)

Reply:
> Cloning im::HashMap is not expensive which is kinda the whole point for using it over std HashMap?

oh, you are right. I somehow (embarrassingly) missed that it is an `im` and not a `std` hashmap. I switched to `im` hashmap myself https://github.com/solana-labs/solana/pull/23585 and I am well aware that clones are cheap.

However the spirit of the comment stays, because we really do not need to "consume" the struct anyways.

> frankly it feels like you're asking me to do a change for some imaginary code that will never exist.

Why is it imaginary?
`stake_delegations` field used to be just a std hash-map. Then https://github.com/solana-labs/solana/pull/21542 wrapped it into an `Arc`. Then https://github.com/solana-labs/solana/pull/23585 switched to `im` hashmap. So it has gone through few changes impacting efficiency of `clone`, and might change again in the future.
In fact for `Stakes<Delegation>` I think it does not need to be an `im` hashmap and just a std hashmap is sufficient and probably faster. (ie. only `Stakes<StakeAccount>` needs to be `im` hashmap).


Anyways, I don't think this part of the change has any performance impact or necessary for this PR.
So lets keep `stakes: &Stakes<Delegation>` argument a reference, and we can discuss in separate smaller PR if changing that arg is an improvement or not.

Reply:
Agreed, I changed it because it seems more correct to me, not for perf reasons. We seem to disagree, so happy to take it out of this PR 🤝

Reply:
ok, then can you please update the pr

Reply:
done